internal/handler/common/update.go

```diff
@@ -31,29 +33,173 @@ import (
 	"github.com/FerretDB/FerretDB/internal/util/must"
 )
 
-// UpdateDocument updates the given document with a series of update operators.
-// Returns true if document was changed.
-// To validate update document, must call ValidateUpdateOperators before calling UpdateDocument.
-// UpdateDocument returns CommandError for findAndModify case-insensitive command name,
-// WriteError for other commands.
-// TODO https://github.com/FerretDB/FerretDB/issues/3013
-func UpdateDocument(command string, doc, update *types.Document, insert bool) (bool, error) {
-	var docUpdated bool
-	var err error
+// UpdateDocument iterates through documents from iter and processes them sequentially based on param.
+// Returns UpdateResult if all operations (update/upsert) are successful.
+//
+// In case of updating multiple documents, UpdateDocument returns an error immediately after one of the
+// operation fails. The rest of the documents are not processed.
+// TODO https://github.com/FerretDB/FerretDB/issues/2612
+func UpdateDocument(ctx context.Context, c backends.Collection, cmd string, iter types.DocumentsIterator, param *Update) (*UpdateResult, error) { //nolint:lll // for readability
+	result := new(UpdateResult)
 
-	if update.Len() == 0 {
-		// replace to empty doc
-		for _, key := range doc.Keys() {
-			docUpdated = true
+	docs, err := iterator.ConsumeValues(iter)
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	result.Matched.Count = int32(len(docs))
+
+	var upsert bool
+
+	if len(docs) == 0 {
+		if !param.Upsert {
+			return result, nil
+		}
+
+		upsert = true
+		docs = append(docs, must.NotFail(types.NewDocument()))
+	}
+
+	for _, doc := range docs {
+		var modified bool
+
+		result.Matched.Doc = doc.DeepCopy()
```

Comment:
`result.Matched.Doc` and `result.Modified.Doc` - both of these pointers are only utilized in case of `findAndModify`. In case of `update` flow the caller func don't use these pointers, just ignores them and only make use the matched/modified counts.

In `findAndModify`, we have a necessity to return either matched doc or modified doc based on the `ReturnNew` flag. That's why we need these pointers. Also, since we are wrapping the document iterator with limit(1) in the caller func, there will be only one matched document, not multiple.